Feat/graph heatmap

FarisElghlan opened the pull request:
Added a heatmap to the graph.

Fixes #228

FarisElghlan committed:
Added an algorithm chooser

It is now possible to select which bubbling algorithms
have to be performed on the graph, see settings menu.

FarisElghlan committed:
Added mutations filters for graph collapsion

FarisElghlan committed:
Set the most frequent lineage color for each edge

Fixes #183

Will  make an option to highlight lower frequency
lineages later if i have enough time.

wouterjsmit committed:
Tidy up JavaDoc and code layout

Adds details to javadoc and moves some lines for better overview.

wouterjsmit committed:
Add further javadoc newlining

FarisElghlan committed:
Added heatmaps to the graph

Fixes #228

wouterjsmit commented:
Please rebase against `dev`

FarisElghlan committed:
Merge branch 'dev' into feat/graph-heatmap

Conflicts:
	PL2/PL2-gui/src/main/java/nl/tudelft/pl2016gr2/gui/view/graph/GraphPaneController.java
	PL2/PL2-gui/src/main/resources/style.css

FarisElghlan commented:
Merged dev into this branch.

wouterjsmit commented:
All commits except the last are already part of dev. Could you in one way or another remove those ocmmits from this history?

FarisElghlan commented:
Is that really important? they aren't shown in the diff anymore and I don't really know how to remove them (besides removing this branch, creating a new branch and cherry picking the changes).

wouterjsmit commented:
Since I'm curious we'll see if Git will actually filter them

wouterjsmit commented on PL2/PL2-gui/src/main/java/nl/tudelft/pl2016gr2/gui/view/graph/GraphPaneController.java:
```diff
@@ -156,10 +182,50 @@ public void initialize(URL location, ResourceBundle resources) {
+   */
+  private void addHeatmapOptions() {
+    GraphicsContext heatmapGraphics = heatmap.getGraphicsContext2D();
+    heatmapOptions.put("Amount of mutations (parralel paths in the graph)",
```
`parralel -> parallel`

wouterjsmit commented on PL2/PL2-gui/src/main/java/nl/tudelft/pl2016gr2/gui/view/graph/heatmap/AbstractHeatmap.java:
```diff
@@ -0,0 +1,120 @@
+
+  @Override
+  public void visit(GraphNode node) {
+  }
```
Do insert something in terms of logging/ assertions to verify that this base version never gets called.

wouterjsmit commented on PL2/PL2-gui/src/main/java/nl/tudelft/pl2016gr2/gui/view/graph/heatmap/AbstractHeatmap.java:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  @Override
+  public void visit(GraphBubble bubble) {
```
The visits are identical for every bubble type. Why differentiate?

FarisElghlan replied:
They are overwritten in the subclasses. This way indel bubble only has to overwrite 1 method for example:

``` java
public class IndelDensity extends AbstractHeatmap {

  public IndelDensity(GraphicsContext heatmapGraphics, ObservableDoubleValue zoomFactor) {
    super(heatmapGraphics, zoomFactor);
  }

  @Override
  public void visit(IndelBubble bubble) {
    setColor(MUTATION_COLOR);
  }

  @Override
  protected double getWidthMultiplier() {
    return 2.0;
  }
}
```

wouterjsmit replied:
But you could simply not define them at all in the abstract class then right? At runtime the correct heatmap is selected, so the overloaded method is available.

FarisElghlan replied:
No, all of the subclasses overwrite a different method, for example the `PhyloBubbleDensity` class overwrites the `void visit(PhyloBubble bubble)` method, while the `PointMutationDensity` class overwrites the `void visit(PointMutationBubble bubble)` method. All methods which are not overriden in the subclasses need to behave as specified in the parent class.

wouterjsmit replied:
Ah, I see your point. Then I agree

FarisElghlan committed:
Fixed 2 edge bugs (#243)

Fixes #233

Bug 1: Now when bubbles are unpopped, the nested bubbles are
forced to be unpopped as well.
Bug 2: The y coordinate of the edge origin and destination aren't
swapped anymore when the origin isn't drawn in the view.

FarisElghlan committed:
Updated according to feedback
